### DISCUSSION TITLE
add `opacity` to composite

### DISCUSSION THREAD

**Discussion by jcupitt** [libvips/libvips#2063] [Ideas]
`composite` needs an opacity parameter.

See https://github.com/libvips/libvips/issues/1622

**Comment by jcupitt** [comment]
Background: when wrirting composite, I think we imagined opacity would be implemented by scaling alpha before adding images to the stack.

However, this is not how layer stacks in programs like PhotoShop actually work. They apply opacity after the blend -- there's an implicit scale + over after each layer. This means that you can't use the libvips many-image composite operator to duplicate the effect of PhotoShop layer stacks and you are forced to do a series of pairwise composite operations plus an etxtra "over" each time.

If we add an `opacity` parameter to `composite` we could compute the extra `over` operation as part of the main composite loop.

**Comment by www-YM** [comment]
Hi @jcupitt ,

I found this package to composite with opacity to control the intensity of blending, like photoshop. But this operation takes a few seconds to composite two images, it's too long to composite two images.

https://github.com/flrs/blend_modes

Is this realized in libvips?

Thanks.

**Comment by jcupitt** [comment]
Hi @www-YM,

Yes, libvips implements all the usual blend modes. It should be reasonably quick. Speed depends on the size and format of the images, of course, and on how libvips was compiled. On this PC I see:

```
$ time vips composite2 wtc.jpg biglion.png x.jpg over --x 1000 --y 1000
real	0m0.992s
user	0m2.888s
sys	0m0.101s
```

That's rendering a 4,000 x 4,000 pixel PNG on top of a 10,000 by 10,000 pixel JPG, and includes all IO, compression, decompression, and so on.

If you see poor performance, please open a new issue with some sample code and images which show the problem.
